aten/src/ATen/native/cuda/MemoryAccess.cuh

```diff
@@ -0,0 +1,253 @@
+template <> struct has_builtin_vector_type<int64_t> : public std::true_type {};
+template <> struct has_builtin_vector_type<float>   : public std::true_type {};
+template <> struct has_builtin_vector_type<double>  : public std::true_type {};
+
+// for types that does not have corresponding builtin vector type,
+// it is ensured that dynamic dispatch will never use it. But
+// we need to create a stub for it for completeness
+template <typename scalar_t>
+struct fake_vector {
+  scalar_t x, y, z, w;
+};
+
+template <typename scalar_t, int size>
+struct Info {
+  static constexpr int alignment = 0;
+  using vector_type = fake_vector<scalar_t>;
+};
+
+template <typename scalar_t>
+struct Info<scalar_t, 1> {
+  static constexpr int alignment = std::alignment_of<scalar_t>::value;
+  using vector_type = scalar_t;
+};
+
+#define DEFINE_VECTOR_INFO(TYPE, SIZE, VECTYPE, ALIGNMENT)    \
+  template <>                                                 \
+  struct Info<TYPE, SIZE> {                                   \
+    static constexpr int alignment = ALIGNMENT;               \
+    using vector_type = VECTYPE;                              \
+  }
+
+// Note: alignment data could be found at:
+// https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#built-in-vector-types
+
+//                    TYPE, SIZE, VECTYPE, ALIGNMENT
+DEFINE_VECTOR_INFO(uint8_t,    2,  uchar2,         2);
+DEFINE_VECTOR_INFO(uint8_t,    4,  uchar4,         4);
+
+DEFINE_VECTOR_INFO( int8_t,    2,   char2,         2);
+DEFINE_VECTOR_INFO( int8_t,    4,   char4,         4);
+
+DEFINE_VECTOR_INFO(int16_t,    2,  short2,         4);
+DEFINE_VECTOR_INFO(int16_t,    4,  short4,         8);
+
+DEFINE_VECTOR_INFO(    int,    2,    int2,         8);
+DEFINE_VECTOR_INFO(    int,    4,    int4,        16);
+
+static_assert(sizeof(int64_t) == sizeof(long long), "long long is assumed to be 64bit");
+DEFINE_VECTOR_INFO(int64_t,    2,   longlong2,        16);
+DEFINE_VECTOR_INFO(int64_t,    4,   longlong4,        16);
+
+DEFINE_VECTOR_INFO(  float,    2,  float2,         8);
+DEFINE_VECTOR_INFO(  float,    4,  float4,        16);
+
+DEFINE_VECTOR_INFO( double,    2, double2,        16);
+DEFINE_VECTOR_INFO( double,    4, double4,        16);
+
+#undef DEFINE_VECTOR_INFO
+
+template <typename scalar_t, int size>
+struct Vec {
+  typename Info<scalar_t, size>::vector_type v;
+  static_assert(size == 1 || size == 2 || size == 4, "only vectors of size 1, 2, 4 are supported");
+};
+
+template <typename scalar_t>
+struct Vec<scalar_t, 1> {
+  typename Info<scalar_t, 1>::vector_type v;
+  __device__ inline scalar_t get(int i) {
+    return v;  // no boundary check here
+  }
+  __device__ inline void set(int i, scalar_t value) {
+    v = value;  // no boundary check here
+  }
+};
+
+template <typename scalar_t>
+struct Vec<scalar_t, 2> {
+  typename Info<scalar_t, 2>::vector_type v;
+  __device__ inline scalar_t get(int i) {
```

Comment:
unfortunately windows and rocm does not allow me to overload `operator[]` here...